[SPIRV][NFC] Refactor pointer creation in GlobalRegistery

s-perron opened the pull request:
This PR adds new interfaces to create pointer type, and adds
some requirements to the old interfaces. This is the first step in
https://github.com/llvm/llvm-project/issues/134119.

s-perron committed:
[SPIRV][NFC] Refactor pointer creation in GlobalRegistery

This PR adds new interfaces to create pointer type, and adds
some requirements to the old interfaces. This is the first step in
https://github.com/llvm/llvm-project/issues/134119.

Keenuts commented on llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.cpp:
```diff
@@ -1693,6 +1728,43 @@ SPIRVType *SPIRVGlobalRegistry::getOrCreateSPIRVArrayType(
 }
 
 SPIRVType *SPIRVGlobalRegistry::getOrCreateSPIRVPointerType(
+    Type *BaseType, MachineInstr &I, SPIRV::StorageClass::StorageClass SC) {
+  MachineIRBuilder MIRBuilder(I);
+  return getOrCreateSPIRVPointerType(BaseType, MIRBuilder, SC);
+}
+
+SPIRVType *SPIRVGlobalRegistry::getOrCreateSPIRVPointerType(
+    Type *BaseType, MachineIRBuilder &MIRBuilder,
```
Those `Type*` can be marked `const`, which will allow you to remove the `const_cast`

s-perron replied:
Good catch. I believe I did not add the const because the original function with the SPIRVType did not use const.

I still have to leave on const cast, but this is a little better.

Keenuts commented on llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.cpp:
```diff
@@ -1693,6 +1728,43 @@ SPIRVType *SPIRVGlobalRegistry::getOrCreateSPIRVArrayType(
+  return R;
+}
+
+SPIRVType *SPIRVGlobalRegistry::getOrCreateSPIRVPointerTypeInternal(
```
Since this is making a lookup in the CurMF, how does this handles functions referencing global types? Are those re-created and each function has its own type version?

s-perron commented on llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.cpp:
```diff
@@ -54,6 +54,40 @@ static unsigned typeToAddressSpace(const Type *Ty) {
+  case SPIRV::StorageClass::Uniform:
+  case SPIRV::StorageClass::PushConstant:
+  case SPIRV::StorageClass::StorageBuffer:
+    // case SPIRV::StorageClass::PhysicalStorageBuffer:
```
It does not exist yet in the backend. I will remove it.

s-perron replied:
It is there with and `EXT`. I've fixed it.

s-perron commented on llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.h:
```diff
@@ -466,6 +466,14 @@ class SPIRVGlobalRegistry : public SPIRVIRMapping {
+  // example, it has the correct offset and stride decorations.
+  SPIRVType *getOrCreateSPIRVPointerTypeInternal(
+      SPIRVType *BaseType, MachineIRBuilder &MIRBuilder,
+      SPIRV::StorageClass::StorageClass SClass = SPIRV::StorageClass::Function);
```
Done

s-perron commented on llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.h:
```diff
@@ -540,13 +548,32 @@ class SPIRVGlobalRegistry : public SPIRVIRMapping {
 
+  // Returns a pointer to a SPIR-V pointer type with the given base type and
+  // storage class. It is the responsibility of the caller to make sure the
+  // decorations on the base type are valid for the given storage class. For
+  // example, it has the correct offset and stride decorations.
```
Yes it is true. The above function take a `Type` and will generate the correct spir-v type for the storage class.

In this function, if you give the wrong SPIRVType, the other functions will end up generating a different SPIRVType for the pointee. If the caller expects that `BaseType` is the pointee type used then there could be an error. This is why I have the assert in the body of this function. If the pointee type used does not match the given SPIR-V type, then the caller violated this condition.

s-perron commented on llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.h:
```diff
@@ -540,13 +548,32 @@ class SPIRVGlobalRegistry : public SPIRVIRMapping {
-      SPIRVType *BaseType, MachineIRBuilder &MIRBuilder,
+      Type *BaseType, MachineIRBuilder &MIRBuilder,
       SPIRV::StorageClass::StorageClass SClass = SPIRV::StorageClass::Function);
   SPIRVType *getOrCreateSPIRVPointerType(
-      SPIRVType *BaseType, MachineInstr &I, const SPIRVInstrInfo &TII,
+      Type *BaseType, MachineInstr &I,
       SPIRV::StorageClass::StorageClass SClass = SPIRV::StorageClass::Function);
 
+  // Returns a pointer to a SPIR-V pointer type with the given base type and
+  // storage class. It is the responsibility of the caller to make sure the
+  // decorations on the base type are valid for the given storage class. For
+  // example, it has the correct offset and stride decorations.
+  SPIRVType *getOrCreateSPIRVPointerType(
+      SPIRVType *BaseType, MachineIRBuilder &MIRBuilder,
+      SPIRV::StorageClass::StorageClass SClass = SPIRV::StorageClass::Function);
```
Done

s-perron commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -3366,18 +3370,20 @@ bool SPIRVInstructionSelector::selectImageWriteIntrinsic(
-    SPIRVType *PtrType =
-        GR.getOrCreateSPIRVPointerType(ResType, MIRBuilder, SC);
+    SPIRVType *PtrType = GR.getOrCreateSPIRVPointerType(
+        const_cast<Type *>(ResType), MIRBuilder, SC);
```
Done.

s-perron committed:
Fixes from code review

Make some paramters const, to remove need for const_cast.

Fix PhysicalStorageBuffer storage class use.

thurstond commented:
> LLVM Buildbot has detected a new failure on builder `ppc64le-lld-multistage-test` running on `ppc64le-lld-multistage-test` while building `llvm` at step 12 "build-stage2-unified-tree".
> 
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/168/builds/10735
> 
> Here is the relevant piece of the build log for the reference

Fixed-forward with a tiny change (remove default case) in https://github.com/llvm/llvm-project/commit/83f831d46f46e9f68bec9b7a6592f678c8ba61ab :-)
